361 D4HJ Update layer key design and include the new unregistered land layer

ms0ur1s opened the pull request:
### What? Why?
This introduces changes to the Layer Key menu to reflect the designs completed in #341. The working branch was also rebased to development to reflect changes implemented in #360. Updates include refactoring actions for toggling property displays, ensuring only one ownership layer is active at a time, and enhancing the styling of buttons, modals, and making the zoom warning at button.

Relates to #361

#### Functional Changes to Land Ownership and Map Layers

* **Refactored `togglePropertyDisplay` Action**: Replaced the old implementation to ensure only one ownership layer is active at a time. The new logic toggles the active display state and updates the Redux store accordingly. (`src/actions/LandOwnershipActions.js`, [src/actions/LandOwnershipActions.jsR108-R128](diffhunk://#diff-21a0cf0568636f02d11f1ebf85a82a93e3aeee974cf33fbdb6e0e55d5a82a8f3R108-R128))
* **New `toggleOwnershipLayerInKey` Action**: Introduced a utility to manage ownership layers in the map key, ensuring only one layer is active and others are toggled off. (`src/actions/MapActions.js`, [src/actions/MapActions.jsR438-R485](diffhunk://#diff-fa33e15812a97a7267658783f33519442036ab8e1afb6a0e64665ed0944f00c5R438-R485))
* **Enhanced Map Initialization**: Added logic to ensure ownership layers are reflected in the menu key upon map load or creation. (`src/actions/MapActions.js`, [[1]](diffhunk://#diff-fa33e15812a97a7267658783f33519442036ab8e1afb6a0e64665ed0944f00c5R113-R128) [[2]](diffhunk://#diff-fa33e15812a97a7267658783f33519442036ab8e1afb6a0e64665ed0944f00c5R172-R174)

#### UI/UX Enhancements

* **Updated Button Styles**: Added a new `.zoom-warning-button` and `.button-clear` styles for better visual feedback and customization. (`src/assets/styles/_buttons.scss`, [[1]](diffhunk://#diff-9983b00b09438d4fbeb991c71e1aeff3ad3b35a6d54c380a469f6e19426c6d1bR162-R197) [[2]](diffhunk://#diff-9983b00b09438d4fbeb991c71e1aeff3ad3b35a6d54c380a469f6e19426c6d1bR344-R356)
* **Improved Tooltip Design**: Redesigned the tooltip for the layer key with better responsiveness and styling. (`src/assets/styles/_tooltips.scss`, [[1]](diffhunk://#diff-93a440eb279e5d85288becab49020fb4a438a7e10aeffb59f1681136258281a5L143-R298) [[2]](diffhunk://#diff-93a440eb279e5d85288becab49020fb4a438a7e10aeffb59f1681136258281a5L315)
* **Refined Modal Close Button**: Added a dark-grey variant for the modal close button for improved accessibility. (`src/assets/styles/_modals.scss`, [[1]](diffhunk://#diff-55edf4afbaee976c843a276eedf2cf9cd6262fd9df4950ebe240c7fdc9b0bca7L9-L12) [[2]](diffhunk://#diff-55edf4afbaee976c843a276eedf2cf9cd6262fd9df4950ebe240c7fdc9b0bca7R21-R28)

#### Component Refactoring

* **Enhanced `LandDataLayerToggle` Component**: Added support for custom toggle handlers and improved the default toggle logic to align with the new Redux actions. (`src/components/left-pane/LandDataLayerToggle.js`, [src/components/left-pane/LandDataLayerToggle.jsL1-R37](diffhunk://#diff-ceb66affb4f92381d88ea25febdc5efa055e47f0da18e64758ef22886b699efbL1-R37))
* **Updated `LeftPaneLandData` Component**: Integrated the new toggle logic for ownership layers and replaced `LeftPaneToggle` with `LandDataLayerToggle` for better modularity. (`src/components/left-pane/LeftPaneLandData.js`, [src/components/left-pane/LeftPaneLandData.jsL122-R168](diffhunk://#diff-11cc046ed1b048cecf2b673027a7096e77486f3ae65a445180ed60cd28c304e2L122-R168))

### What should we test?
- Open the **Data Layer** panel
- Toggle a layer from the **Land Data** group
- Click the red "Zoom in to see layer" button
- Once zoomed in toggle a layer from the **Land Ownership** group
- Test selecting and clearing properties and check that selected properties are still visible when zooming out or disabling the layer.
- Create a new map and ensure all **Data Layers** are reset
- Test between different saved map
- Check mobile UX

ms0ur1s committed:
Refactor land ownership display logic and enhance layer management

- Updated `togglePropertyDisplay` to improve state handling and logging.
- Added new ownership layer colors in `_variables.scss`.
- Enhanced `LandDataLayerToggle` to support custom toggle functions.
- Integrated ownership toggle handling in `LeftPaneLandData`.
- Improved layer visibility management in `MenuKey`.
- Updated reducers to handle new layer actions and state updates.

ms0ur1s committed:
Refactor land ownership display logic and enhance layer management

- Updated `togglePropertyDisplay` to improve state handling and logging.
- Added new ownership layer colors in `_variables.scss`.
- Enhanced `LandDataLayerToggle` to support custom toggle functions.
- Integrated ownership toggle handling in `LeftPaneLandData`.
- Improved layer visibility management in `MenuKey`.
- Updated reducers to handle new layer actions and state updates.

ms0ur1s committed:
Ownership layers displayed in menu key after map load

ms0ur1s committed:
Ownership layers displayed in menu key after map load

ms0ur1s committed:
Ensure ownership key toggles correctly

Only one ownership key visible at a time

ms0ur1s committed:
Ensure ownership key toggles correctly

Only one ownership key visible at a time

ms0ur1s committed:
Ownership key visible at property boundry zoom level

ms0ur1s committed:
Ownership key visible at property boundry zoom level

ms0ur1s committed:
Added ticket number to comments

to keep track of changes

ms0ur1s committed:
Added ticket number to comments

to keep track of changes

ms0ur1s committed:
Initial layer key styling

ms0ur1s committed:
Initial layer key styling

ms0ur1s committed:
update desktop layer key styles

ms0ur1s committed:
update desktop layer key styles

ms0ur1s committed:
360 d4hj add new colours to land ownership layers (#364)

Co-authored-by: John Evans <38507954+King-Mob@users.noreply.github.com>

ms0ur1s committed:
Styling update - layer key header, content and tab

ms0ur1s committed:
Styling update - layer key header, content and tab

ms0ur1s committed:
Added animation to the layer menu

ms0ur1s committed:
Added animation to the layer menu

ms0ur1s committed:
Move Layer Key to MapboxMap

from ControlButtons.js, minor reduction in feedback button size

ms0ur1s committed:
Move Layer Key to MapboxMap

from ControlButtons.js, minor reduction in feedback button size

ms0ur1s committed:
Update ZoomWarning to control Layer Menu Visibility

ms0ur1s committed:
Update ZoomWarning to control Layer Menu Visibility

ms0ur1s committed:
Layer Key display - only hidden at ownership level

ms0ur1s committed:
Layer Key display - only hidden at ownership level

ms0ur1s committed:
MenuKey visibility condition updated

ms0ur1s committed:
MenuKey visibility condition updated

ms0ur1s committed:
ZoomWarning converted to zoom button

ms0ur1s committed:
ZoomWarning converted to zoom button

ms0ur1s committed:
MenuKey - added higlighted properties

ms0ur1s committed:
MenuKey - added higlighted properties

ms0ur1s committed:
MenuKey - landownership key colour updates

ms0ur1s committed:
MenuKey - landownership key colour updates

ms0ur1s committed:
MenuKey, Key - include fill & border properties

ms0ur1s committed:
MenuKey, Key - include fill & border properties

ms0ur1s committed:
MenuKey - update border style for active property to dashed

ms0ur1s committed:
MenuKey - update border style for active property to dashed

ms0ur1s committed:
Mobile key and button style updates

ms0ur1s committed:
Mobile key and button style updates

ms0ur1s committed:
MenuKey - mobile menu handling with animation

ms0ur1s committed:
MenuKey - mobile menu handling with animation

ms0ur1s committed:
Code clean up

ms0ur1s committed:
Code clean up

ms0ur1s committed:
Merge branch '361-d4hj-update-layer-key-design-and-include-land-ownership-layers' of https://github.com/DigitalCommons/land-explorer-front-end into 361-d4hj-update-layer-key-design-and-include-land-ownership-layers

ms0ur1s committed:
Add clearMapLayers action & reducer  case

to reset map layers state

rogup reviewed (commented):
Hey @ms0ur1s 
The new UI and colours are looking nice when I gave it a little test in my dev environment. However, there seemed to be lots of bugs, with the key sometimes disappearing altogether when I changed the layers.

Ultimately, I think the bugs are due to the logic throughout not being properly thought out. New code was added in lots of places where there was already existing code doing a similar thing. Looking at the commit history, it seems that more and more duplicate code was added to plug gaps in a whack-a-mole kind of way, similar to the collaborative mapping work that we did a while ago.
I'm sorry to say this and it feels harsh because a lot of time has clearly been spent already, but most of this logic needs a re-write. Because it's going to be extremely difficult to debug the issues, and it's impacting our codebase quality so that working on this part of the code in the future will be difficult.

As a general rule of thumb before writing the code, we always need to have a clear understanding of where the state (in this case, using Redux) will be defined, ensuring it's as concise as possible and we don't keep multiple sources of the same truth. We also need to read around the existing code to understand what's already there and what we can work with. Then we need to decide what flows of logic need to be added or changed to achieve our acceptance criteria, in the simplest way possible.

I'm thinking that we need to formalise this process more, so that the code design can be reviewed and refined before lots of time is spent writing new code.

For this PR, I'm not sure what the best next step is, since it depends on how much resource we have (@lin-d-hop ?) . I suppose you could have a go at writing up a design doc that includes details about
- where the state for map layers and the key currently lives in the dev branch, and any changes that you propose making to this
- a list of changes to the UI components
- a list of the logic changes that need to be made in order to display the ownership layer state in the key

Then once I've reviewed that, you can have a go at another PR. A lot of your UI work could probably be copied over since that part is looking good.

ms0ur1s committed:
First resolutions

Revert togglePropertyDisplay and remove clearMapLayers action

ms0ur1s committed:
Second resolutions

Simplified visibility logic in MenuKey and MapboxMap components

ms0ur1s commented on src/reducers/MapLayersReducer.js:
```diff
@@ -5,23 +5,99 @@ let landDataLayers;
+        const toggleLayerId = action.payload;
+
+        /**
+         * #361 - Handle toggling of land data layers
```
Will remove at the end. These were just for my convenience and I forgot to remove them.

ms0ur1s committed:
Third resolutions

Refactor MenuKey and MapboxMap to remove unecessary state

ms0ur1s committed:
missing import added after testing

ms0ur1s committed:
Reverted to previous logic

Reverted to using landDataLayers and propertiesDisplay for MenuKey condition

Removed unecessary logic from MapActions

ms0ur1s committed:
removed complex display logic

display conditions handled in MapboxMap

ms0ur1s commented on src/components/left-pane/LeftPaneLandData.js:
```diff
@@ -66,6 +71,23 @@ const LeftPaneLandData = ({ open, active, onClose }) => {
   );
 
+  // #361 - Handle toggling ownership layers
+  const handleOwnershipToggle = (display) => {
```
Reverted to using `LeftPaneToggle` with its original action and reverted `LandDataLayerToggle` to its previous state.

ms0ur1s committed:
Removed duplicate toggle actions

Removed handleOwnershipToggle, reverted to LeftPaneToggle for Land Ownersip layers and reverted LandDataLayerToggle to its previous state

ms0ur1s commented on src/actions/MapActions.js:
```diff
@@ -415,4 +434,52 @@ const shortenTimestamp = (timestamp) => {
+ * #361 - Toggle ownership layer in the key
+ * Ensures only one ownership layer is active in the key at a time.
+ */
+export const toggleOwnershipLayerInKey = (layerId) => {
```
Removed

ms0ur1s committed:
clear highlighted properties on openMap

Currently highlighed properties are cleared when new map is loaded

ms0ur1s commented on src/actions/MapActions.js:
```diff
@@ -109,7 +111,24 @@ export const openMap = (mapId) => {
+      //     type: "ENSURE_LAYER_IN_KEY",
+      //     payload: layerId,
+      //   });
+      // }
```
Removed

ms0ur1s commented on src/actions/MapActions.js:
```diff
@@ -110,6 +110,22 @@ export const openMap = (mapId) => {
+
+        // Ensure the layer is in the key
+        dispatch({
+          type: "ENSURE_LAYER_IN_KEY",
```
Removed

ms0ur1s committed:
Removing unused code & comments

ms0ur1s committed:
Rename  MapLayersReducer & TOGGLE_LAYER

Removed LR_POLYGONS_ENABLED and all references

ms0ur1s committed:
Rename MenuKey to MapLayerKey

and centre zoom button

ms0ur1s committed:
MapBeingEditedToast centred

ms0ur1s committed:
LandDataLayersReducer reverted to dev version

ms0ur1s committed:
MapLayerKey component refactor and cleanup

ms0ur1s committed:
MapLayerKey > moved keys to external config

ms0ur1s committed:
Refactor getKeys for correct layer order

ms0ur1s committed:
update conditionals in MapboxMap

Conditional check for displaying MapLayerKey and removal of logs

rogup committed:
Zoom button fixes

rogup committed:
Tidy and fix layer key

rogup committed:
Fix MapProperties so that highlighted proeprties are always visible

rogup committed:
File renaming

rogup committed:
Add unregistered and pending layers + fix how Mapbox layers work

rogup committed:
Prefer using reducer rather than chain of actions

rogup commented:
Hey @ms0ur1s
I've added a few commits which resolve the above comments, and there were also quite a few other changes to fix other things such as the bugginess of the multiple Mapbox layers (see commit messages for a summary, or ask me if you'd like more context)

I'd usually do a rebase to try to neaten the commits and then merge, but this would be very complex in this case, and I think the priority is to get this merged. So I'm gonna do a squash merge now :)
